atc/worker/gardenruntime/worker.go

```diff
@@ -713,3 +640,97 @@ func (worker *Worker) GardenClient() gclient.Client {
 func (worker *Worker) BaggageclaimClient() baggageclaim.Client {
 	return worker.bcClient
 }
+
+func (worker *Worker) acquireVolumeStreamingLock(
+	ctx context.Context,
+	resourceCacheID int,
+	destination string,
+) (lock.Lock, bool, error) {
+	// if caching of streamed volumes is not enabled then we should not globally
+	// lock the volume streaming since steps cannot share the streamed volumes as caches
+	if !atc.EnableCacheStreamedVolumes {
+		return lock.NoopLock{}, true, nil
+	}
+
+	logger := lagerctx.FromContext(ctx)
+	return worker.db.LockFactory.Acquire(logger, lock.NewVolumeStreamingLockID(resourceCacheID, destination))
+}
+
+func (worker *Worker) findOrStreamVolume(
+	ctx context.Context,
+	privileged bool,
+	teamID int,
+	container db.CreatingContainer,
+	artifact runtime.Artifact,
+	mountPath string,
+	delegate runtime.BuildStepDelegate,
+) (Volume, error) {
+	logger := lagerctx.FromContext(ctx)
+	var tick *time.Ticker
+
+	_, inputPath := path.Split(mountPath)
+	if inputPath == "" {
+		inputPath = "for image"
+	}
+
+	for {
+		localVolume, ok, err := worker.findVolumeForArtifact(ctx, teamID, artifact)
+		if err != nil {
+			logger.Error("failed-to-find-volume-for-artifact", err)
+			return Volume{}, err
+		}
+		if ok && localVolume.DBVolume().WorkerName() == worker.Name() {
+			volume := localVolume.(Volume)
+			return volume, nil
+		}
+
+		volume, ok := artifact.(Volume)
+		var resourceCacheLock lock.Lock
+		var acquired bool
+		if ok {
+			resourceCacheLock, acquired, err = worker.acquireVolumeStreamingLock(ctx, volume.dbVolume.GetResourceCacheID(), container.WorkerName())
+		} else {
+			// If the artifact is not a volume we cannot cache it so always stream
+			resourceCacheLock, acquired = lock.NoopLock{}, true
+		}
+		if err != nil {
+			logger.Error("failed-to-acquire-lock-for-volume-streaming", err)
+			return Volume{}, err
+		}
+
+		if acquired {
+			defer resourceCacheLock.Release()
+
+			streamedVolume, err := worker.findOrCreateVolumeForStreaming(
+				ctx,
+				privileged,
+				container,
+				teamID,
+				mountPath,
+			)
+			if err != nil {
+				logger.Error("failed-to-create-image-artifact-replicated-volume", err)
+				return Volume{}, err
+			}
+
+			delegate.StreamingVolume(logger, inputPath, artifact.Source(), streamedVolume.DBVolume().WorkerName())
+			if err := worker.streamer.Stream(ctx, artifact, streamedVolume); err != nil {
+				logger.Error("failed-to-stream-artifact", err)
+				return Volume{}, err
+			}
+			logger.Debug("streamed-non-local-volume")
```

Comment:
Add some data like streamedVolume handle, dest worker name.

Reply:
This is all logged by the streamer already - not sure if we really need it again here? https://github.com/concourse/concourse/blob/master/atc/worker/streamer.go#L41-L46

Reply:
ok, then just ignore this comment.